### PR TITLE
[READY] Fix overmap special spawning

### PR DESCRIPTION
**Pull request by kevingranade**
Guarantees that minimum special counts are satisfied by created overmap or a nearby one.

### CODE REVIEW

**Comment by kevingranade** [comment]
Still has a fairly critical issue, in that it seems to be completely incapable of spawning starting locations now :/

**Comment by Coolthulhu** [comment]
Any plans about that edge case which I mentioned on the forums?
When no overmap can generate a specific special, it will not be generated no matter how many times is it passed over to neighbors.
And locking city settings to prevent that would be worse than the problem it tries to fix.

**Comment by DangerNoodle** [comment]
> Still has a fairly critical issue, in that it seems to be completely incapable of spawning starting locations now :/

That is impressive, given I had recently just fixed several starting locations and other specials that previously failed to spawn.

**Comment by kevingranade** [comment]
I can't think of a more general solution to that problem other than making overmap generation and special placement a LOT more restrictive.
The unit tests will now catch that case (as long as said special has a minimum value) for default configuration.  Overmap generation needs to be a LOT faster before we can afford to repeat this test for e.g. different city size settings, so detecting it in extreme configurations will have to wait for either some holistic speedup or a workaround like generating a single overmap with each test configuration, and then copying it and re-running different special placement tests on each one.

**Comment by kevingranade** [comment]
Found the issue, I misunderstood the documentation for constructor delegaion.

**Comment by Leland** [comment]
Experiencing #21135 again with this PR – try starting on the Out of Town scenario. Also, not seeing beehives spawn, which was fixed by #21136.

**Comment by DangerNoodle** [comment]
Regressions are undesirable.

**Comment by kevingranade** [comment]
I'll look into alternate starting locations, possibly they're spawning on something other than the origin overlap and that's causing issues.

@DangerNoodle neither of your comments on this issue are productive, if you don't have specific feedback or suggestions, please refrain from posting.

**Comment by DangerNoodle** [comment]
My apologies.

**Comment by kevingranade** [comment]
Adressed some nagging issues regarding test fragility and ability of system to initialize missions and starting locations.
Tested starting location spawning by repeatedly (~6 times) creating new worlds and creating a new character with the Out of Town scenario, chosing a mix of starting locations.
Tested mission spawn by visiting the shelter console and looking up the refugee center.  I got it to repeatedly find refugee centers on different overmaps.

File: tests/overmap_test.cpp [hunk 1]
```diff
@@ -19,3 +20,21 @@ TEST_CASE( "set_and_get_overmap_scents" ) {
...
+
+TEST_CASE( "default_overmap_generation_always_succeeds" ) {
+    // List of specials to generate, overmap should consume them all.
+    std::vector<const overmap_special *> default_specials = overmap( 0, 0 ).get_enabled_specials();
+    for( int i = 1; i < 10; i++ ) {
+        std::vector<overmap_special_placement> test_specials;
+        for( auto special : default_specials ) {
+            test_specials.push_back( { 0, special } );
+        }
+        overmap_buffer.create_custom_overmap( i, i, test_specials );
```
Comment (Coolthulhu) [inline tests/overmap_test.cpp]: Only diagonals? That's an easy case.
You use nearest points in radius when deferring specials. Using that here would be more likely to catch problems.

File: src/overmap.cpp [hunk 2]
```diff
@@ -4244,95 +4227,63 @@ void overmap::place_specials_pass( std::vector<std::pair<const overmap_special *
...
+    // First insure that all minimum instance counts are met.
+    place_specials_pass( enabled_specials, sectors, false );
+    if( std::any_of( enabled_specials.begin(), enabled_specials.end(),
+                     []( overmap_special_placement placement ) {
+                         return placement.instances_placed < placement.special_details->occurrences.min;
+                     } ) ) {
+        // Randomly select from among the nearest uninitialized overmap positions.
+        int previous_distance = 0;
+        std::vector<point> nearest_candidates;
+        for( point candidate_addr : closest_points_first( 10, pos() ) ) {
```
Comment (Coolthulhu) [inline src/overmap.cpp]: 10 overmap radius? That's risky because the errors will never be shown, even when the nearest 100 maps is all bugged.

File: src/overmap.cpp [hunk 3]
```diff
@@ -4244,95 +4227,63 @@ void overmap::place_specials_pass( std::vector<std::pair<const overmap_special *
...
+                } else {
+                    break;
+                }
+            }
+        }
+        if( !nearest_candidates.empty() ) {
+            std::random_shuffle( nearest_candidates.begin(), nearest_candidates.end() );
+            point new_om_addr = nearest_candidates.front();
+            overmap_buffer.create_custom_overmap( new_om_addr.x, new_om_addr.y, enabled_specials );
+        }
```
Comment (Coolthulhu) [inline src/overmap.cpp]: There is that mostly-unused `dbg(x)` function. It's suboptimal in that it only seems to work in some builds (debug?), but it's better to use that than to just leave it all silent and sweep the bugs under the rug.

File: src/overmap.cpp (hunk 2 above)
Comment (kevingranade) [inline src/overmap.cpp]: You have evidence that it breaks?

Comment (kevingranade) [inline src/overmap.cpp]: i.e. how do you end up with a 20x20 square of initialized overmaps that don't have appropriate specials?  Remember the goal here is to insure a certain distribution of specials, not for each invocation to complete.

File: src/overmap.cpp (hunk 3 above)
Comment (kevingranade) [inline src/overmap.cpp]: Considering that the algorithm painting itself into a corner is not a bug, how do you propose we detect a failure?  For that matter, what is a faiure here?

File: tests/overmap_test.cpp (hunk 1 above)
Comment (kevingranade) [inline tests/overmap_test.cpp]: Sure, done.

File: src/overmap.cpp (hunk 2 above)
Comment (Coolthulhu) [inline src/overmap.cpp]: I don't have evidence that it breaks, but it will be hard to gather it if it avoids displaying the errors until the last moment (ie. searching for mission).
We will have to ask people to make sure their overmaps aren't migrated from old worldgen, then there will be a painful process of reproducing the error with no real tools for it.
Then, when new specials are added, there will be no surefire method of checking if they don't break, only the probabilistic unit test.

Then there are mods and non-standard settings, which aren't covered under unit tests at all, meaning they are at huge risk of breaking with no good info.

File: tests/overmap_test.cpp [hunk 4]
```diff
@@ -19,3 +21,30 @@ TEST_CASE( "set_and_get_overmap_scents" ) {
...
+    for( point candidate_addr : closest_points_first( 10, { 0, 0 } ) ) {
+         // Skip populated overmaps.
+      if( overmap_buffer.has( candidate_addr.x, candidate_addr.y ) ) {
+            continue;
+        }
+        std::vector<overmap_special_placement> test_specials;
+        for( auto special : default_specials ) {
+            test_specials.push_back( { 0, special } );
+        }
+        overmap_buffer.create_custom_overmap( candidate_addr.x, candidate_addr.y, test_specials );
```
Comment (Coolthulhu) [inline tests/overmap_test.cpp]: I'm not sure if I understand what this function does.
Does it overwrite the specials or not?
Reading the function, it looks to me like it drops the whole overmap and generates a new one, with the specials from argument. If so, it doesn't really test that the specials are passed around properly and don't "overflow".

File: src/overmap.cpp (hunk 3 above)
Comment (Coolthulhu) [inline src/overmap.cpp]: A failure would be not generating a mandatory special in some range (3x3 block of overmaps?).
The problem at hand is that sometimes missions will not get an object they require. If the overmap generation can't 100% guarantee this object exists, this problem will not be fixed. And that should be reported as soon as possible.

So one way to fix it would be to keep the deferred specials in "must generate right here" vector. Or flag them as must-generate.

A better solution would be to somehow keep the special data in the overmap buffer and allow generating the specials around the place where they failed to generate. That is, on failure, move them to a point nearest the original failed overmap, not to a neighbor of current overmap (with baggage of deferred specials).
After failing to generate a given special in such a block, it should be reported to the player, even if by a debugmsg. Because it is a bug that will result in other bugs (mission errors).

Basically, there must be some sort of method to ensure that every NxN-sized block of overmaps has at least one special of every mandatory type.

Comment (kevingranade) [inline src/overmap.cpp]: > A failure would be not generating a mandatory special in some range (3x3 block of overmaps?).

Do you have a way to check for that?

> So one way to fix it would be to keep the deferred specials in "must generate right here" vector. Or flag them as must-generate.

Must generate where? The whole point of this issue is that we don't have a way to consistently generate the full set of specials on a given overmap. They're already flagged as "must generate" by being marked mandatory.

File: tests/overmap_test.cpp (hunk 4 above)
Comment (kevingranade) [inline tests/overmap_test.cpp]: Each iteration through this loop is essentially independent, they just happen in the same game world because generating a new one is super slow.  That having been said, this code excersises the logic for scanning for non-existent overlaps and uncovered some bugs related to that.  In pseudocode:

```
Until we've placed 10 sets of specials:
    Find an adjacent and uninitialized overmap location
        Construct an overmap there using a full list of specials
            Recursively find nearby overmap positions
                generate overmaps until mandatory specials are all placed.
        Assert that list of specials has been exausted of mandatory specials.
```
More tersely, the specials list isn't reset until we've successfully asserted that it generated all of its specials.

File: src/overmap.cpp (hunk 3 above)
Comment (kevingranade) [inline src/overmap.cpp]: The problem with checking for placed specials is that the overmap generation code doesn't know what they look like, we could possibly add a list of overmap terrain types to each special object, then overmap generation could assert that it has either generated each mandatory special OR that it exists in range.

Comment (Coolthulhu) [inline src/overmap.cpp]: "Must generate or print an error"
If an object fails to generate within 2 overmap radius from the overmap where it was supposed to be, it must print a warning.

It is a big regression if there is no such warning.
It's not acceptable to generate all overmaps until it fails. Is there a proper end condition or will it only fail when stack depth is exceeded or writing a file fails? I'm not sure if I'm reading this right, but it looks to me like it will recursively push the excess specials around for as long as it can.
If it keeps generating overmaps indefinitely, there is no way of checking if the settings/specials are good enough to actually ever produce an overmap.

It would wreck the ability to test extensive mods (say, More Locations, that ships with core) and also make it much harder to generate a world with "too many" (hard to tell what is "too many" without manually testing) mods that add new specials.
Even worse if we add custom settings, which don't even warn that the world gen might fail.

I see the possible solutions:
* Pass around stack depth (as in, number of `generate_custom_overmap` above it) to the generating function. Fail if it exceeds some value, preferably configurable, so that players with weaker computers don't wait all day to get that failure message.
* With each special, pass the overmap coord of the overmap where this special was supposed to spawn (first such overmap). If current failed overmapgen is >2 overmap radius from any special's parent overmap, fail generation.

Also, it would be more "robust" if the `generate_custom_overmap` tried to cluster the new overmaps around the starting point instead of always going in one direction (up, I think).

Comment (kevingranade) [inline src/overmap.cpp]: > If an object fails to generate within 2 overmap radius from the overmap where it was supposed to be, it must print a warning.

This is incorrect behavior in the rather mundane scenario of a player walking in a large circle and then exploring the center, at some point it will fail to generate additional overmaps due to all available locations being populated, but in reality there will already be sufficient specials in range.

> If it keeps generating overmaps indefinitely, there is no way of checking if the settings/specials are good enough to actually ever produce an overmap.

We can easily terminate recursion and warn on failure to place even a single mandatory special in an overmap, but that isn't a full solution.

> Also, it would be more "robust" if the generate_custom_overmap tried to cluster the new overmaps around the starting point instead of always going in one direction (up, I think).

It doesn't, reread overmap.cpp lines 4260-4283

None of your proposals recognize that overmap generation can and will fail occasionally due to exhaustion of nearby overmaps, and that this is not a bug.  If you don't get this, this discussion is utterly unproductive.  Also I'm far less concerned with theoretical bugs caused by misbehaving mods than I am by the status quo, which is that users are presented with a dialog asking them to retry mapgen indefinitely or accept a broken game world.

I'm inclined to adjust special definitions to include their terrain types and cut the generation range, but allow pre-existing specials to count toward the requisite totals, and warn on failure to either generate or find the required specials, but not in this pr.

Comment (Coolthulhu) [inline src/overmap.cpp]: > This is incorrect behavior in the rather mundane scenario of a player walking in a large circle and then exploring the center, at some point it will fail to generate additional overmaps due to all available locations being populated, but in reality there will already be sufficient specials in range.

Except when it isn't the case. You don't guarantee that it happens anywhere. It's perfectly possible that all the surrounding overmaps also failed to generate that special.
> We can easily terminate recursion and warn on failure to place even a single mandatory special in an overmap, but that isn't a full solution.

Neither is pushing the specials away and allowing them to be silently skipped on slightly unusual settings.

> None of your proposals recognize that overmap generation can and will fail occasionally due to exhaustion of nearby overmaps, and that this is not a bug.

You described the PR as:
> Guarantees that minimum special counts are satisfied by created overmap or a nearby one.

Which one is it then? Guarantees or can fail?
You don't check neighbors of current overmap to make sure that the specials can be pushed away or already have the special it needs.
> Also I'm far less concerned with theoretical bugs caused by misbehaving mods than I am by the status quo, which is that users are presented with a dialog asking them to retry mapgen indefinitely or accept a broken game world.

Your PR does not prevent broken worlds from being generated.
If the dialog is a problem, it's literally one string change to make it default to accepting the broken worlds silently. They aren't significantly more broken than the ones you generate in this PR, they just have higher chance (just chance) to lack desired density of specials.
Killing the ugly dialog is not worth massive regression that is locking up on perfectly OK settings, without warning or option to disable it. The dialog can be disabled in the options, meaning it is much less intrusive than infinite loop.
Infinite loop without warning on allowed settings is worse than just plain crashing.

Comment (kevingranade) [inline src/overmap.cpp]: > make it default to accepting the broken worlds silently. They aren't significantly more broken than the ones you generate in this PR, 

The broken worlds the old code generated were missing an arbitrary subset of the mandatory specials when using defaults, in the same scenario the new code generates all mandatory specials. The scenarios you've outlined that would cause the new scheme to fail (total inability to generate a particular special, excessive mandatory specials causing resource exhaustion) would similarly break the old scheme, and with the addition of warning on inability to generate a single special on an overmap, the first would fail just as visibly, and the second would always be more successful at generating more usable specials.

> is locking up on perfectly OK settings,

Please provide those settings.

Comment (Coolthulhu) [inline src/overmap.cpp]: > Please provide those settings.

Maximum city size, minimum city distance.
This setting makes it nearly impossible to generate some specials. It's fine if it just informs the player of failure, it's not fine if it tries to generate them forever or silently fails to generate them.

Comment (kevingranade) [inline src/overmap.cpp]: This should fail the, "can't generate any specials" test, which I've said I'm adding.

Comment (kevingranade) [inline src/overmap.cpp]: Results of testing with those settings:
Previous to this PR: Completely incapable of generating a valid map, shows a menu and demands user accepts broken map.
With this PR: Took a very long time, but eventually terminated after generating 1032 overmaps.
Latest version with check for failure to place any specials returns immediately with a failure message.

File: src/start_location.cpp [hunk 5]
```diff
@@ -199,44 +199,20 @@ void start_location::prepare_map( tinymap &m ) const
...
-                _( "The game could not create a world with a suitable starting location.\n\n"
-                   "Depending on the world options, the starting location may never appear. If the problem persists, you can try another starting location, or change the world options.\n\n"
-                   "Try again?" ) ) ) {
-            return overmap::invalid_tripoint;
+            return omtstart + point( omp.x * OMAPX, omp.y * OMAPY );
         }
-        overmap_buffer.clear();
     }
+    // Should never happen, if it does we messed up.
+    popup( _( "Unable to generate a valid starting location, please report this falure." ) );
```
Comment (Coolthulhu) [inline src/start_location.cpp]: Typo: "failure"

File: tests/test_main.cpp [hunk 6]
```diff
@@ -79,17 +81,20 @@ void init_global_game_state( const std::vector<std::string> &mods )
...
     g->load_world_modfiles( world_generator->active_world );
 
     g->u = player();
-    g->u.create(PLTYPE_NOW);
+    g->u.create( PLTYPE_NOW );
 
     g->m = map( get_option<bool>( "ZLEVELS" ) );
 
+    std::vector<overmap_special_placement> empty_specials;
+    overmap_buffer.create_custom_overmap( g->get_levx(), g->get_levy(), empty_specials );
```
Comment (Coolthulhu) [inline tests/test_main.cpp]: I'm pretty sure those x/y arguments are of wrong "type": `create_custom_overmap` passes its arguments to overmap constructor, meaning those should be in overmap coords (as in, one coord - one overmap).
But `get_levx/y` are in submap coords (as in, one coord, one submap).

**Comment by Coolthulhu** [comment]
The test doesn't seem to test anything at the moment, I'm not getting any output even with the `-s` switch.

File: src/overmap.h [hunk 7]
```diff
@@ -444,32 +439,37 @@ class overmap
...
+    void place_special( const overmap_special &special, const tripoint &p, om_direction::type dir, const city &cit );
+    /**
+     * Iterate over the sectors of the overmap and place the quota of specials.
+     * If the stated minimums are not reached, it will spawn a new nearby overmap
+     * and continue placing specials there.
+     * @param enabled_specials specifies what specials to place, and tracks how many have been placed.
+     **/
+    void place_specials( std::vector<overmap_special_placement> &enabled_specials );
+    /**
+     * One pass of placing specials - by default there are 3 (mandatory, mandatory without city distance, optional)
```
Comment (Coolthulhu) [inline src/overmap.h]: Are there still 3 passes?

**Comment by Coolthulhu** [comment]
Currently doesn't compile due to implicit casting away of constness of reference at `overmap.cpp:4234`.

Other than that the above, it seems to be working.

**Comment by Coolthulhu** [comment]
I'll be "stress testing" it with unofficial mods now.
Meanwhile it would be very useful to have the following:
* A test case that actually tests if all the specials are spawned in some block (3x3 or 5x5), so that placement of mission objectives is actually tested
* Listing to debug log all the specials pushed to a different overmap. Otherwise it is nearly impossible to track what overmaps are pushed around without debug breakpoints
* A hard test case, possibly marked as can-fail (preferably not). Such as the city size one

I did some manual testing, but manual testing is always a lower standard.

**Comment by Coolthulhu** [comment]
Well, that could be why it started generating new overmaps indefinitely in a hard to generate world. It would probably keep going until it crashed due to lack of memory.

That's a giant regression that can't wait for the next PR.

**Comment by kevingranade** [comment]
"Limit special placement recursion" eliminates the potential for indefinite recursion (it wasn't actually infinite, it did a random walk and seemed to terminate fairly reliably, but it was still unacceptable behavior as it took 10+ min to happen).

The starting point for overmap placement is persisted, and only overmaps within distance 2 are eligible for generation, meaning a 5x5 grid, or 25 overmaps created maximum.  On my system this takes about 10s, but it should do for a failsafe until I add a system that guarantees "generation or discovery" of specials within the same range.  I tried dialing it down to distance 1, but bandit camps would sometimes not generate in that range, probably because in some cases half or more of the overmaps were populated already, meaning it had only ~3-4 overmaps to attempt.

Very occasional warnings from this system are expected, for example if a player explores the borders  of a 5x5 overmap sized area and then explores the center.  This may result in an isolated overmap being generated that is unable to recurse to nearby neighbors.  Note the warning in this case is almost certainly a false alarm, as the overmap would be surrounded on all sides by already-generated overmaps that do meet the criteria of special reachability.

Regarding reachability, the system guarantees either a warning or a full complement of specials within range 2 of the origin overmap, and reflexively guarantees a full complement of specials within a range of 4 of any generated overmap (overmaps generated at maximum distance from the origin overmap in opposite directions are distance 4 from each other).  As such the code that searches for specials searches an area with a radius of 4 inclusive, guaranteeing discovery of a mandatory special if not preceeded by a warning.

The next step would be to add code to search existing overmaps in range for qualifying specials, which would stand in for newly generated specials.  This would reduce the occurrence of spurrious warnings, and possibly allow the reduction of the generation range limit for overmaps to 1.  This will be a significant chunk of code, and is likely to present some odd corner cases, so I would prefer not to incorporate it into this PR.

**Comment by Coolthulhu** [comment]
Needs a rebase.

Would be useful to have one test of the hard settings, such as max city size, minimum distance.
Not strictly necessary if the new code is 100% guaranteed (no RNG) to terminate within sane time.

Regarding the false alarms:
It would be very useful to have overmaps care about generated neighbors when pushing around specials.
For example, like this:
* Each overmap stores a map of generated specials
* If overmap is generated next to an existing overmap and it checks it, it doesn't just ignore it as unavailable slot, but accounts for the specials on that overmap
* Accounting for specials can be, for example, treating them as if they were newly generated specials (decrementing the required specials in overmap batch)
* (Optional) Before pushing around any specials, the overmap could check its neighborhood for those specials. If they are found, there would be no need to push any of them to new overmaps. Neighborhood checks before pushing could be done in circles (distance 1, distance 2 etc.) to avoid overly sparse (in specials) overmaps - first check all neighbors in distance 1, if that fails push specials in distance 1, then check neighbors in distance 2 and so on

This would handle the false alarms and slightly decrease order dependence on generation. Order dependence in worldgen is a minor flaw, as the world should not care how it is uncovered.

The above suggestion is not necessary here, but I'm pretty sure it would be very simple to implement.

**Comment by kevingranade** [comment]
> Would be useful to have one test of the hard settings, such as max city size, minimum distance.

From two weeks ago:
>  Overmap generation needs to be a LOT faster before we can afford to repeat this test for e.g. different city size settings, so detecting it in extreme configurations will have to wait for either some holistic speedup or a workaround like generating a single overmap with each test configuration, and then copying it and re-running different special placement tests on each one.


> It would be very useful to have overmaps care about generated neighbors when pushing around specials.
> [plus details]

From a week ago:
> The  next step would be to add code to search existing overmaps in range for  qualifying specials, which would stand in for newly generated specials.   This would reduce the occurrence of spurrious warnings, and possibly  allow the reduction of the generation range limit for overmaps to 1.   This will be a significant chunk of code, and is likely to present some  odd corner cases, so I would prefer not to incorporate it into this PR.
